DESIGN.md

```diff
@@ -0,0 +1,773 @@
+
+```rust
+fn main() {
+    let rt = tokio_uring::runtime::Runtime::new().unwrap();
+    rt.block_on(async {
+        // The rest of the application comes here.
+    });
+}
+```
+
+The application's `main` function starts a tokio-uring runtime and launches its
+asynchronous logic within that. The tokio-uring runtime can drive both io-uring
+specific resources (e.g., `TcpStream` and `TcpListener`) and Tokio resources,
+enabling any library written for Tokio to run within the io-uring specific
+runtime.
+
+## Submit-based operations
+
+Operations on io-uring backed resources return futures, representing the
+operation completion. The caller awaits the future to get the operation result.
+
+```rust
+let socket = my_listener.accept().await?;
+```
+
+The runtime communicates with the kernel using two single-producer,
+single-consumer queues. It submits operation requests, such as accepting a TCP
+socket, to the kernel using the submission queue. The kernel then performs the
+operation. On completion, the kernel returns the operation results via the
+completion queue and notifies the process. The `io_uring_enter` syscall flushes
+the submission queue and acquires any pending completion events. Upon request,
+this syscall may block the thread waiting for a minimum number of completion
+events. Both queues use memory shared between the process and the kernel and
+synchronize with atomic operations.
+
+Operation futures provide an asynchronous cancel function, enabling the caller
+to await on a clean cancellation.
+
+```rust
+let accept = tcp_listener.accept();
+
+tokio::select! {
+    (tcp_stream, addr) = &mut accept => { ... }
+    _ = tokio::time::sleep(Duration::from_millis(100)) => {
+        // Accept timed out, cancel the in-flight accept.
+        match accept.cancel().await {
+            Ok(_) => { ... } // operation canceled gracefully
+            Err(Completed(tcp_stream, addr)) => {
+                // The operation completed between the timeout
+                // and cancellation request.
+            }
+            Err(_) => { ... }
+        }
+    }
+}
+```
+
+In practice, operation timeouts will deserve a dedicated API to handle the
+boilerplate.
+
+```rust
+let (tcp_stream, addr) = tcp_listener
+    .accept()
+    .timeout(Duration::from_millis(100))
+    .await?
+```
+
+The cancel and timeout function are inherent methods on operation future types.
+
+If the operation's future drops before the operation completes, the runtime will
+submit a cancellation request for the operation. However, cancellation is
+asynchronous and best-effort, the operation may still complete. In that case,
+the runtime discards the operation result.
+
+The queue's single-producer characteristic optimizes for a single thread to own
+a given submission queue. Supporting a multi-threaded runtime requires either
+synchronizing pushing onto the submission queue or creating a queue pair per
+thread. The tokio-uring runtime uses an isolated thread-per-core model,
```

Comment:
Does 'thread-per-core' model here imply CPU affinity as well? I wonder `sched_setaffinity()` will be planned or not as I could not see related implementation code yet.